--- RTSim-1.1.2.c	2010-09-23 06:57:06.000000000 +0200
+++ RTSim.c	2020-01-08 22:28:28.351005488 +0100
@@ -329,9 +329,11 @@
   for (i = 0; i < work->num_input_streams; i++) {
     FREE(work->input_component_name[i]);
   }
+  FREE(work->input_component_name);
   for (i = 0; i < work->num_output_streams; i++) {
     FREE(work->output_component_name[i]);
   }
+  FREE(work->output_component_name);
 
   FREE(work->input_stream);
   FREE(work->output_stream);
